include/deal.II/base/parameter_acceptor.h

```diff
@@ -0,0 +1,523 @@
+ * @endcode
+ *
+ * If instead one of the paths ends with "/" instead of just
+ * a name of the class, subsequent classes will interpret this as
+ * a full path, interpreting the class name as a directory name:
+ * @code
+ * int main()
+ * {
+ *   MyClass ca("/Class A/Class/");
+ *   MyClass cb("Class B/Class");
+ *   ParameterAcceptor::initialize("file.prm");
+ * }
+ * @endcode
+ * The parameter file will have the following structure
+ * @code
+ * subsection Class A
+ *   subsection Class
+ *      ...
+ *      subsection Forcing term
+ *      ...
+ *      end
+ *      subsection Class B
+ *          subsection Class
+ *          ...
+ *          end
+ *          subsection Forcing term
+ *          ...
+ *          end
+ *      end
+ *   end
+ * end
+ * @endcode
+ *
+ * As a final remark, in order to allow a proper management of all the
+ * sections/subsections, the instantiation of objects and the call to
+ * ParameterAcceptor::initialize() cannot be done on multiple, concurrently
+ * running threads.
+ *
+ * If you pass an empty name, the boost::core::demangle() function is used to
+ * fill the section name with a human readable version of the class name
+ * itself.
+ *
+ * @author Luca Heltai, 2017.
+ */
+class ParameterAcceptor : public Subscriptor
+{
+public:
+  /**
+   * The constructor adds derived classes to the list of acceptors. If
+   * a section name is specified, then this is used to scope the
+   * parameters in the given section, otherwise a pretty printed
+   * version of the derived class is used.
+   */
+  ParameterAcceptor(const std::string &section_name="");
+
+  /**
+   * Destructor.
+   */
+  virtual ~ParameterAcceptor();
+
+  /**
+   * Call declare_all_parameters(), read the parameters from `filename` (only
+   * if `filename` is a non-empty string), and then call
+   * parse_all_parameters().
+   *
+   * If the parameter `filename` is the empty string, then no attempt to read a
+   * parameter file is done. This may be useful if you are ok with using
+   * default values, and don't want to read external files to use a class
+   * derived from ParameterAcceptor.
+   *
+   * If outfilename is not the emtpy string, then write the content that was
+   * read in to the outfilename. The format of both input and output files are
+   * selected using the extensions of the files themselves. This can be either
+   * `prm` or `xml` for input, and `prm`, `xml`, or `tex/latex` for output. If
+   * the output format is `prm`, then `output_style_for_prm_format` is used to
+   * decide wether we write the full documentation as well, or only the
+   * parameters.
+   *
+   * If the input file does not exist, a default one with the same name is created
+   * for you, and an exception is thrown.
+   *
+   * @param filename Input file name
+   * @param output_filename Output file name
+   * @param output_style_for_prm_format How to write the output file if format is `prm`
+   */
+  static void initialize(const std::string &filename="",
```

Comment:
i have one suggestion: I think it's worth adding another `initialize` with `std::istream` (and `ostream`) and move the code there. For  unit tests in user libraries reading parameters from a stream is super handy when your test must be `.cc` and not just a `.prm`.

Reply:
A stream is also very useful if you want to do some pre-processing before you pass the contents of the file to `ParameterHandler`. We do this in ASPECT, for example.

Reply:
Done. I also added an optional argument to the other initialize function, so that one can use an already existing prm instead of the global `ParameterAcceptor::prm`.